sycl/include/sycl/device_aspect_macros.hpp

```diff
@@ -0,0 +1,389 @@
+//==------------------- device_aspect_macros.hpp - SYCL device -------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#ifndef __SYCL_ALL_DEVICES_HAVE_0__
+// __SYCL_ASPECT(host, 0)
+#define __SYCL_ALL_DEVICES_HAVE_0__ 0
+#endif
+
+#ifndef __SYCL_ALL_DEVICES_HAVE_1__
+// __SYCL_ASPECT(cpu, 1)
+#define __SYCL_ALL_DEVICES_HAVE_1__ 0
+#endif
+
+#ifndef __SYCL_ALL_DEVICES_HAVE_2__
+//__SYCL_ASPECT(gpu, 2)
+#define __SYCL_ALL_DEVICES_HAVE_2__ 0
+#endif
+
+#ifndef __SYCL_ALL_DEVICES_HAVE_3__
+//__SYCL_ASPECT(accelerator, 3)
+#define __SYCL_ALL_DEVICES_HAVE_3__ 0
+#endif
+
+#ifndef __SYCL_ALL_DEVICES_HAVE_4__
+//__SYCL_ASPECT(custom, 4
+#define __SYCL_ALL_DEVICES_HAVE_4__ 0
+#endif
+
+#ifndef __SYCL_ALL_DEVICES_HAVE_5__
+// __SYCL_ASPECT(fp16, 5)
+#define __SYCL_ALL_DEVICES_HAVE_5__ 0
+#endif
+
+#ifndef __SYCL_ALL_DEVICES_HAVE_6__
+// __SYCL_ASPECT(fp64, 6)
+#define __SYCL_ALL_DEVICES_HAVE_6__ 0
+#endif
+
+#ifndef __SYCL_ALL_DEVICES_HAVE_9__
+// __SYCL_ASPECT(image, 9
+#define __SYCL_ALL_DEVICES_HAVE_9__ 0
+#endif
```

Comment:
Technically, we have aspects `7` and `8`, they are documented in a separate file:

But they are still part of the `enum class aspect`:
https://github.com/intel/llvm/blob/6761f0ef372a0e54b970d85241808339e9bcf077/sycl/include/sycl/info/aspects_deprecated.def#L1-L4

https://github.com/intel/llvm/blob/6761f0ef372a0e54b970d85241808339e9bcf077/sycl/include/sycl/aspects.hpp#L15-L23

I was wondering: do we want to craft some python script (or something like that) to auto-generate this and `device_aspect_traits.hpp` files? Apparently, it will take some time now, but perhaps it will help with further maintenance?

Reply:
@steffenlarsen is the responsible for the design, so I think he should voice his opinion too. IMO, it would be a bigger burden to maintain the python script than this. The test will make sure that every aspect has a corresponding device aspect trait, and the developer that added a new aspect simply has to copypaste and slightly adapt a few lines. Like I said, I think creating and maintaining the Python script is more costly than maintaining this manually.

Reply:
Either should work and though the Python script would mean less work when adding a new aspect, I worry it would obfuscate the actual structure of the implementation as it may not be obvious that something like this would be generated by some odd script run at compile-time.

Reply:
Good point about maintenance and initial cost: I suppose most of the work is to create an initial version of this header, but subsequent updates should be easy to do. I will mark this as resolved then